### DISCUSSION TITLE
how to receive binary package?

### DISCUSSION THREAD

**Discussion by evehaxor** [miguelgrinberg/python-socketio#730] [Q&A]
Hello, How do I interpret binary packages using the socket.io module?

![binary](https://user-images.githubusercontent.com/73003571/112357418-45e89280-8cae-11eb-8ec6-01c1fd531668.jpg)
![binary2](https://user-images.githubusercontent.com/73003571/112357691-88aa6a80-8cae-11eb-99d3-ce9b3a4185c4.png)

python-engineio  3.14.2
python-socketio  4.6.0

**Comment by miguelgrinberg** [comment]
Binary messages are passed on to handlers in the same way as regular JSON messages. Are you familiar with the Socket.IO protocol?

**Comment by evehaxor** [comment]
> Binary messages are passed on to handlers in the same way as regular JSON messages. Are you familiar with the Socket.IO protocol?

I'm learning, can you give me an example?

**Comment by miguelgrinberg** [comment]
@evehaxor there is no difference between binary and non-binary data. Just use a bytes object for your binary packets. You can send just a binary payload, or you can combine JSON with binary, both work.

**Comment by evehaxor** [comment]
I don't know how to interpret binary messages in python

```
@sio.on('s_balance/list')
def message(data):
    print(data)
```
    
it works

but how do I do binary messages? thanks for listening

**Comment by miguelgrinberg** [comment]
There are no changes to your code for binary data. The `data` argument would be of type `bytes` if the other side sends a binary payload.

**Comment by fcy197778770** [comment]
> Binary messages are passed on to handlers in the same way as regular JSON messages. Are you familiar with the Socket.IO protocol?

Emitting event "message" [/stock]
Sending packet MESSAGE data 51-/stock,["message",{"_placeholder":true,"num":0}]
Sending packet MESSAGE data <binary>

hi，what does  {"_placeholder":true,"num":0}  mean？？ 
I don't understand，

sio.emit(event='trdCmd', data=b'\x08\x05\x10h*\x1b\x12\x0860000011\x1a\x06123456J\x071.7.7.7', namespace="/stock")

this is mye code and  i can't Received anything

**Comment by miguelgrinberg** [comment]
@fcy197778770 the placeholder is where the binary data goes. This data is sent in a follow up package. Read the socket.io protocol documentation if you are interested in the internals.

**Comment by fcy197778770** [comment]
> @fcy197778770 the placeholder is where the binary data goes. This data is sent in a follow up package. Read the socket.io protocol documentation if you are interested in the internals.

@miguelgrinberg  Hi,  I checked the log，I did send sessages and the server received a message.
but It looks like a mess。 
![image](https://user-images.githubusercontent.com/68221995/122165876-6a1dcf00-ceab-11eb-82c0-861e926a37e5.png)
the ERROR line  is  the message i send.  
It should be b'\x08\xba\xd0\xe7\x04\x10h*\x1c\x12\t880331110\x1a\x06123456J\x071.7.7.7:!27e9c10d246e63971a6680f2eba6a8fb5'

**Comment by miguelgrinberg** [comment]
@fcy197778770 This is not a log that comes from this package, so I don't know how to interpret it. Please create a small server and client application that I can use to reproduce the problem here.
